Document text index commands (#1247)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/ql/administration/indexes-for-search-performance/index.asciidoc

Once an index has been created, it will be managed and kept up to date by the DBMS.
Neo4j will automatically pick up and start using the index once it has been created and brought online.

> There are multiple index types available, b-tree, fulltext, lookup, and text index types.

Reviewer: ```suggestion
There are multiple index types available: b-tree, fulltext, lookup, and text index types.
```

> More information about fulltext indexes are found in <<administration-indexes-fulltext-search>>.

Reviewer: ```suggestion
See <<administration-indexes-fulltext-search>> for more information about fulltext indexes .
```

> Token lookup indexes contains nodes with one or more label or relationship type, without any regard for properties.

Reviewer: ```suggestion
Token lookup indexes contain nodes with one or more labels or relationship types, without regard for any properties.
```

> Cypher enables the creation of b-tree indexes on one or more properties for all nodes or relationships that have a given label or relationship type:

Reviewer: ```suggestion
Cypher enables the creation of b-tree indexes on one or more properties for all nodes or relationships with a given label or relationship type:
```

* An index that is created on a single property for any given label or relationship type is called a _single-property index_.
* An index that is created on more than one property for any given label or relationship type is called a _composite index_.

Differences in the usage patterns between composite and single-property indexes are described in <<administration-indexes-single-vs-composite-index>>.

> Text indexes are also a version of single-property indexes, with the limitation that they only knows about properties with string values.

Reviewer: ```suggestion
Additionally, a text index is a form of single-property index, with the limitation that it only recognizes properties with string values.
```

CREATE LOOKUP INDEX [index_name] [IF NOT EXISTS]
FOR (n)
ON EACH labels(n)

> [OPTIONS "{" option: value[, ...] "}"]

Reviewer 2: Strictly speaking , the `option: value[, ...]` isn't quite correct since there only exists one specific option for lookup and text indexes. Not sure if we want to reflect that here and in the glossary though, or leave it as is and just look at it as a general OPTIONS meta syntax?

Author: You can send in an empty config map as well :P It feels better to have the general syntax here and then specify what is allowed instead of restricting it here (also makes it easier to add other things to it :P)

Reviewer 2: Ok, good point with the empty map. Let's keeps it as it is then

| [source, cypher, role=noplay]
----
CREATE LOOKUP INDEX [index_name] [IF NOT EXISTS]
FOR ()-"["r"]"-()
ON [EACH] type(r)
[OPTIONS "{" option: value[, ...] "}"]
----
| Create a relationship type lookup index.

Index provider can be specified using the `OPTIONS` clause.

| [source, cypher, role=noplay]
----
CREATE TEXT INDEX [index_name] [IF NOT EXISTS]
FOR (n:LabelName)
ON (n.propertyName)
[OPTIONS "{" option: value[, ...] "}"]
----
| Create a text index on nodes.

Index provider can be specified using the `OPTIONS` clause.

| [source, cypher, role=noplay]
----
CREATE TEXT INDEX [index_name] [IF NOT EXISTS]
FOR ()-"["r:TYPE_NAME"]"-()
ON (r.propertyName)
[OPTIONS "{" option: value[, ...] "}"]
----
| Create a text index on relationships.

Index provider can be specified using the `OPTIONS` clause.

| [source, cypher, role=noplay]
----
DROP INDEX index_name [IF EXISTS]
----
| Drop an index of any index type.
| The command is optionally idempotent, with the default behavior to throw an error if you attempt to drop the same index twice.
With `IF EXISTS`, no error is thrown and nothing happens should the index not exist.

| [source, cypher, role=noplay]
----

> SHOW [ALL\|BTREE\|FULLTEXT\|TEXT\|LOOKUP] INDEX[ES]

Reviewer 2: If want them alphabetically, LOOKUP should be before TEXT.

Author: I'm planning on adding RANGE between BTREE and FULLTEXT... :P (since range is the common replacement for btree) and maybe text should be before fulltext as well though

Or I guess I could update to have all in alphabetical order

Reviewer 2: I think either a logical or alphabetical order is ok, this is not really any of it :P

Author: might just be easier to add it in alphabetical, then it takes less thinking to update :P

Index provider can be specified using the `OPTIONS` clause.

| [source, cypher, role=noplay]
----
CREATE TEXT INDEX [index_name] [IF NOT EXISTS]
FOR (n:LabelName)
ON (n.propertyName)
[OPTIONS "{" option: value[, ...] "}"]
----

> | Create a text index on nodes.

Author: Should we mention somewhere that text index only handles properties with string values? And if so, where would be good to add that? @Lojjs

Reviewer 2: That make sense yes, maybe we should extend the index for search performance introduction section `1. Indexes (types and limitations)` with that info?

| `+`

| `type`

> | The IndexType of this index (`BTREE`, `FULLTEXT`, `TEXT`, or `LOOKUP`).

Reviewer 2: In most other places we mention `LOOKUP` before `TEXT`, which also happen to alphabetical order, maybe you want that order here too.

[NOTE]
====
While the command for listing indexes require <<administration-security-administration-database-indexes, the `SHOW INDEX` privilege>>,

> the deprecated built-in procedures for listing indexes, such as `db.indexes`, work as before and are not affected by the privilege.

Reviewer 2: This looks like an unrelated change, which I thought we did already do 🤔 , or was that in some other place/version?

Author: It is higher up in the file, but I did add it here as well when I realized it might be good to have in both places

Reviewer 2: Ok, just thought it looked familiar
